aws/src/main/java/org/apache/iceberg/aws/s3/signer/S3V4RestSignerClient.java

```diff
@@ -0,0 +1,328 @@
+    return Collections::emptyMap;
+  }
+
+  @Value.Lazy
+  public URI baseSignerUri() {
+    return URI.create(
+        properties().getOrDefault(S3_SIGNER_URI, properties().get(CatalogProperties.URI)));
+  }
+
+  @Value.Lazy
+  public String endpoint() {
+    return properties().getOrDefault(S3_SIGNER_ENDPOINT, S3_SIGNER_DEFAULT_ENDPOINT);
+  }
+
+  /** A credential to exchange for a token in the OAuth2 client credentials flow. */
+  @Nullable
+  @Value.Lazy
+  public String credential() {
+    return properties().get(OAuth2Properties.CREDENTIAL);
+  }
+
+  /** A Bearer token which will be used for interaction with the server. */
+  @Nullable
+  @Value.Lazy
+  public String token() {
+    return properties().get(OAuth2Properties.TOKEN);
+  }
+
+  @Value.Lazy
+  RESTClient httpClient() {
+    // TODO: should be closed
+    return HTTPClient.builder()
+        .uri(baseSignerUri().toString())
+        .withObjectMapper(S3ObjectMapper.mapper())
+        .build();
+  }
+
+  @Value.Lazy
+  AuthSession authSession() {
+    AuthSession session = AuthSession.empty();
+    OAuthTokenResponse authResponse = null;
+    if (null != credential() && !credential().isEmpty()) {
+      authResponse = OAuth2Util.fetchToken(httpClient(), session.headers(), credential(), "sign");
+    }
+
+    if (null != authResponse) {
+      session = newSession(authResponse, session);
+    } else if (null != token()) {
+      session = newSession(token(), session);
+    }
+
+    return session;
+  }
+
+  @Value.Check
+  protected void check() {
+    Preconditions.checkArgument(
+        properties().containsKey(S3_SIGNER_URI) || properties().containsKey(CatalogProperties.URI),
+        "S3 signer service URI is required");
+  }
+
+  private Long tokenExpiresInMs() {
+    return PropertyUtil.propertyAsLong(
+        properties(),
+        OAuth2Properties.TOKEN_EXPIRES_IN_MS,
+        OAuth2Properties.TOKEN_EXPIRES_IN_MS_DEFAULT);
+  }
+
+  @Override
+  protected void processRequestPayload(
+      SdkHttpFullRequest.Builder mutableRequest,
+      byte[] signature,
+      byte[] signingKey,
+      Aws4SignerRequestParams signerRequestParams,
+      AwsS3V4SignerParams signerParams) {
+    checkSignerParams(signerParams);
+  }
+
+  @Override
+  protected void processRequestPayload(
+      SdkHttpFullRequest.Builder mutableRequest,
+      byte[] signature,
+      byte[] signingKey,
+      Aws4SignerRequestParams signerRequestParams,
+      AwsS3V4SignerParams signerParams,
+      SdkChecksum sdkChecksum) {
+    checkSignerParams(signerParams);
+  }
+
+  @Override
+  protected String calculateContentHashPresign(
+      SdkHttpFullRequest.Builder mutableRequest, Aws4PresignerParams signerParams) {
+    return UNSIGNED_PAYLOAD;
+  }
+
+  @Override
+  public SdkHttpFullRequest presign(
+      SdkHttpFullRequest request, ExecutionAttributes executionAttributes) {
+    throw new UnsupportedOperationException("Pre-signing not allowed.");
```

Comment:
I think that this isn't supported. We could extend the signer to support pre-signing.

Reply:
is this something we'd want to do as part of this PR or a separate one?

Reply:
because we might need a separate REST endpoint to do pre-signing

Reply:
Pre-signed urls is a very different protocol.  We probably don't want to support that until there's a specific need as it would require a number of changes to how requests are made.